### ISSUE TITLE
Per-gptel-buffer contexts

### ISSUE THREAD

**Issue by orge-dev** [karthink/gptel#475]
**Describe the solution you'd like**
I like to define sets of files/buffers as context, and it would be convenient to have a gptel buffer correspond to a particular set of files as context. So I could have `* Claude (No Context) * buffer, * Claude (Project A1) * buffer, *Claude Project A2* `,etc.

**Describe alternatives you've considered**
An easy way to clear the whole gptel context, plus user defined `(gptel-add-file file)` functions would serve a similar purpose. I'm not sure how to easily remove all items currently in the context.

**Additional context**
Thanks for creating and maintaining gptel!

**Comment by karthink** [comment]
I think I understand what you mean.  Let's call it the "buffer-local context sets" feature, because you can have independent sets of context.  This feature makes sense in the abstract, but not in the details.  Here's why I think it's under-specified:

1. Suppose you visit some buffer, select a region and run `gpel-add` to add it to the context.
   - Which gptel buffer's context (context set) should this be added to?
   - How about when there is no chat buffer?  gptel is usable in any buffer.

2. Same question as 1, but for when you run `gptel-add` in a dired buffer to add a file.

3. Suppose you run `gptel-add-file`, either from the transient menu or via `M-x`.  This command can be run in any buffer.  Which context set should it be added to?

I imagine what you're describing is actually "project-local context sets", not "buffer-local context sets".  Now each project (as managed by project.el or projectile) has an independent gptel context, and any of the above actions 1-3 will add the region/buffer/file to its project's context.  In this case, what happens when you run/want to include a region/buffer/file in project A's context that does not belong to project A?

**Comment by karthink** [comment]
This proposal for multiple context buffers is going to be confusing to most users.  I would prefer `gptel-add` and co to be a one-shot interaction -- every additional decision the user has to make when taking a gptel action will annoy them, as evidenced by past threads like #374 (among others).  If we can find a way to make gptel-add Do The Right Thing in all situations I wouldn't mind working on this.

I like the context link proposal in #481 --it's a very simple mental model that I expect users won't have trouble following.  let's discuss this in the dedicated thread.

**Comment by dwrz** [comment]
I am a new user to gptel -- thank you for all your work on it. I have been really happy with my experience so far, but this issue is the first limitation I felt I ran into. Would it be possible to just have a singular buffer local contexts? 

I may have a chat going on for one project, another for another project, another for a topic, and then I may throw a quick question into *scratch*, and so on. When switching across these, if I need to use context, I have` to dump the context, and then rebuild it -- and in some cases that can be quite a few files. For me, having the context be buffer local would be sufficient, but perhaps I am not thinking about all the use cases -- the only one I can think of is jumping to different files to ask for a refactor / rewrite; you'd have to still rebuild your context.

The only other thing I can think of is being able to assign contexts to variables, and then pull them up by history the same way we can with past directives.

**Comment by karthink** [comment]
@dwrz I feel the pain of dumping and rebuilding the context all the time.  I just don't know how to resolve it.

Buffer-local context does not work because it's ambiguous _which buffer's_ context is being added to when you run `gptel-add`.  

As suggested by @metachip, one way to solve this is to simply ask the user which context set they want to add to.  My experiences with this kind of branching in gptel suggests that it's going to be more annoying than useful, and quite confusing too, leading to doubts like:
- "Wait, which context set is active right now?"
- "Did I add the file to the right context set?"
It also requires gptel to maintain an internal registry of all active context sets, invalidate this registry etc.

I like the idea of using Org/markdown links to add to the context, as discussed in #481, as the context set automatically becomes buffer-local, transparent and easily editable.  But that has other issues, including that you can't have "live" buffer regions in the context.

**Comment by metachip** [comment]
> This proposal for multiple context buffers is going to be confusing to most users. I would prefer gptel-add and co to be a one-shot interaction -- every additional decision the user has to make when taking a gptel action will annoy them, as evidenced by past threads like #374 (among others). If we can find a way to make gptel-add Do The Right Thing in all situations I wouldn't mind working on this.

Agreed. It needs to be DWIM as far as possible.

1) Can you not just use the prefix argument `C-u` ?

   That is, always add/remove context from the global context, if `gptel-add*` called directly or via the transient menu.

   If called directly with a prefix argument, it pops open the list of available context buffers and allows one to create a new one, just like `gptel` does now with respect to chat buffers and their names.

   Existing users will be none the wiser and `gptel-add` remains a one-shot operation.

2) Add another widget to the top of chat buffers to indentify which context is being used.

   This would only show if the context is not the default.

   You could merge it into the context widget you already have, which in addition to reporting the files and regions, says the name of the context too.

@dwrz

> I have` to dump the context, and then rebuild it -- and in some cases that can be quite a few files.

For me this is at the very heart of the problem that we are discussing here.

1) The idea of context, local to a chat buffer as you suggest is one solution.

2) The association of one, of several named context buffers, to a chat buffer, is another. This is my preferred solution.

3) Context that appears in a buffer at the moment, could, for any org mode buffer, be put in a drawer called (say) :CONTEXT: (analogous to the special case of :PROPERTIES: for properties.

All that said, I think links in a buffer may be a better solution per [#481](https://github.com/karthink/gptel/issues/481).

I'll continue my thoughts in that thread.

**Comment by karthink** [comment]
> Can you not just use the prefix argument C-u ?

There is quite a bit more to it than that.  For example, it causes ambiguity in the indication of the current context in the transient menu and in constructing the context buffer (_which_ context set is looked up?).  There's no longer a single source for `*context-buffer*`, compounding the update problems discussed in #482.

I also think you are underestimating the amount of confusion an opaque many-to-one associative map from buffers to context sets will cause when using gptel.  It will be something you have to keep double-checking before sending requests.  In this specific aspect having "project-local" context sets works a little better -- this is also a many-to-one map but people are already used to thinking this way about collections of files/buffers.

**Comment by metachip** [comment]
> I also think you are underestimating the amount of confusion an opaque many-to-one associative map from buffers to context sets will cause when using gptel.

Fair enough. I think links are better anyway because they would be more explicit in terms of the context one is working with.

I have some thoughts about [the questions you have raise about links](https://github.com/karthink/gptel/issues/481#issuecomment-2496450308) which I will answer in that ticket shortly.

**Comment by dwrz** [comment]
@karthink -- thank you! I'm starting to see that there's much more to it than what a first impression suggests.

What about the possibility of using variables to store contexts?

For example:
```
(setq my-context-a '("/home/user/some-file" "/home/user/some-file-2" "/home/user/some-file-3"))
(setq my-context-b '("/etc/foo" "/etc/bar" "/etc/baz"))
(setq gptel-contexts '(my-context-a my-context-b))
```

Then in the transient menu, either let the user build a context as is currently possible, or load one from a submenu, as is possible for directives. Could something like that work?

**Comment by nlevnaut** [comment]
Thinking through the UX for multi-context, perhaps this would work without being too confusing for the default case:
- By default just provide 1 "context group". Existing add/remove commands stays the same.
- Let the user create new named context groups. (not per gptel buffer, just global)
- Switch between the currently-in-use context group (globally) with a `gptel-context-switch` command.

**Comment by karthink** [comment]
> Thinking through the UX for multi-context, perhaps this would work without being
> too confusing for the default case:
> - By default just provide 1 "context group". Existing add/remove commands stays
>   the same.
> - Let the user create new named context groups. (not per gptel buffer, just
>   global)
> - Switch between the currently-in-use context group (globally) with a
>   `gptel-context-switch` command.

Because context state is stored in buffer overlays, looking at buffers will not give you an idea about which regions are included in the current context.  This can get confusing very quickly.

However it should work fine for personal use, if you're aware that the source of truth is what the transient menu (or context inspection buffer) says:
``` emacs-lisp
(defvar gptel-context-groups nil)

(defun gptel-context-switch (group-name)
  "Switch to or create gptel context GROUP-NAME."
  (interactive
   (list (completing-read "Context group (switch or create): "
                          gptel-context-groups)))
  ;; Save current group if required
  (unless gptel-context-groups
    (push (cons "default" gptel-context--alist) gptel-context-groups))
  ;; Switch to or create group
  (if-let* ((group (cdr-safe (assoc group-name gptel-context-groups))))
      (setq gptel-context--alist group)
    (setf (alist-get group-name gptel-context-groups
                     nil nil #'equal)
          (copy-tree gptel-context--alist))))
```

Call `M-x gptel-context-switch` to switch context groups.

**Comment by metachip** [comment]
@ultronozm Very interesting.

The UI you have created to add all sorts of context types looks great. I'll give it a go.

What follows are my initial thoughts in light of what you have done, other things discussed in this tread and some others: Context links (#481), Context management (#484) and Context buffer functionality (#482).

I still have an open mind about the best way to manage context. Buffer local, project local or explicitly named. They all make sense for different use cases. However, there are some challenges when integrating all this into the workflow of GPTel. In particular, ambiguity about which context is active, which context to add to or delete from and how to see what a context comprises. This is especially true because it needs to handle both chat and non-chat buffer scenarios.

I think this is really a UI question more than anything else. I agree with your suggestion about the use of indirect buffers instead of overlays. I don't think it is necessary to try and show all of the context in a single buffer as GPTel tries to do now. Lists of references to files, regions or other entities which when activated open in an indirect for regions or buffers or new buffer if the item is from a file should be enough.

As for knowing which context applies, your solution, making context local to the buffer, is simple and straightforward.

That said, I think a combination of explicitly named contexts, a default context, and means by which one can choose which one to apply at each invocation of gptel-send is better. This is because GPTel is designed to "dance around emacs" rather than remain focused on a single buffer. Chat buffers are the exception of course but I think @karthink is very keen to apply the KISS principle as much as possible, with the lightest of touches on any other mode, function or feature offered by emacs.

I think we can integrate something along the lines of what you have developed, but within a more flexible framework of named contexts, where the context that is applied depends on a set of simple rules which intuitively default back to what GPTel does now.

- At gptel-send time
  - An optional quick context selector
  - Default to buffer-local setting if exists
  - Fall back to global default context
- Persistent association ideas
  - Buffer-local variable to store preferred context
  - In the PROPERTIES drawer (in org mode files)
  - Project-level context preferences

There would also need to be simple means by which context can be managed. We need to create contexts, edit them, add/remove files, regions and metadata. Clear, merge, copy, save and restore.

How one does this presumably means we'd create some context editing mode that opens a context in a buffer and gives us tools to manage it. I have fleshed out a proposal for org mode buffers, using links #481, which extends the metaphor of media links that are already part of GPTel to enable context to be created, edited and deleted by simply embedding org mode links in text.

Anyway, my two cents for now, I will give it some more thought.

**Comment by metachip** [comment]
I forgot to add. I think context should include more than just references to regions, files and media. It should also include what GPTEL already manages; the selected model, the back end, the system prompt.

**Comment by armindarvish** [comment]
@karthink I am going over this discussion and I have a hard time understanding why the `gptel-context--alist` cannot be a buffer-local variable and the single source of truth for context in each buffer (chat buffer or otherwise). 

- What am I missing? Is the challenge with making overlays in buffers? 
If at the end, all you are doing to add context is to append the text to "prompt" or "system" message, then why do you need the overlays? And why can't they be made on the fly from `gptel-context--alist` and not the other way around (making overlays first and keeping a record in the alist)? 

- Is the challenge with sharing the same chat across different buffers (like in the same project)? 
But, `gptel-context--alist` can be looked up in other project-specific buffers and copied to the buffer-local variable (and with some customization switch, the user can optionally be asked to confirm this or not).

- Is the propblem with using `gptel-add`?
Well `gptel-add` can always add the context to the current buffer-local `gptel-context--alist` without keeping track of all different contexts.

**Comment by karthink** [comment]
> @karthink I am going over this discussion and I have a hard time understanding why the gptel-context--alist cannot be a buffer-local variable and the single source of truth for context in each buffer (chat buffer or otherwise).

The problem is that context from other buffers necessarily violates the buffer-local metaphor, so setting `gptel-context--alist` is not the same as setting other parameters like the backend or model per-buffer.

We can work through the consequences:  Let's suppose `gptel-context--alist` is buffer-local.

Suppose also that you have chat buffer (C), with other (code/prose) buffers (A) and (B) germane to your current task.

You also have chat buffer (D), which relates to activity in buffers (E) and (F).  (We need this complication because otherwise the global context works fine.)

You go to buffer (B), select a region and add it to the "context" with `gptel-add`.

Which buffer's value of `gptel-context--alist` should be updated?

> Is the propblem with using gptel-add?
> Well gptel-add can always add the context to the current buffer-local gptel-context--alist without keeping track of all different contexts.

According to this it would be buffer (B)'s `context--alist`.  But this is not useful because we want to use this context when chatting in (C).  How would `gptel-add` know to add this to (C)'s value of `gptel-context--alist`?

Either it should decide automatically (covered below), or it should ask the user.  In both cases you run into state synchronization problems.

> Is the challenge with sharing the same chat across different buffers (like in the same project)?
> But, gptel-context--alist can be looked up in other project-specific buffers and copied to the buffer-local variable (and with some customization switch, the user can optionally be asked to confirm this or not).

According to this we would do something like the following:  when adding a region, buffer or file to the context in a buffer's local value of `gptel-context--alist`, find all buffers belonging to that project, then update `gptel-context--alist` in all of them.

- Now suppose you open some more files from that project, and want to run an operation with gptel (rewriting a region, say) in one of these files.  `gptel-context--alist` in this project buffer has not been updated so we are out of sync.  The only way around this is to add to `find-file-hook` -- only that's not enough!  What about non-file project buffers that might be spawned, like compilation or output buffers?  So we'd have to register a handler in `buffer-list-update-hook` instead.  This is generally a _bad idea_.  This also doesn't cover all buffers, because there are special buffers whose text you may want to add to the context, but don't run `buffer-list-update-hook` when they are spawned/killed.

- Suppose you run `gptel-send` from the transient menu from a utility buffer (like `*Messages*` or `*scratch*`) that's not associated with any project.  The context is now inaccessible.  This problem gets worse when you want to add a non-project (or non-generic-group) file to the project's context from dired, as opposed to from the chat buffer.  The dired directory will not be a project directory.

- It's now a mystery which context set an overlay belongs to, or what the active `context--alist` is at any time.  The UI is ambiguous now -- you can have two buffers side by side with visible context overlays and have no idea if one, both or none of the regions will be sent with your next request.

> What am I missing? Is the challenge with making overlays in buffers?
> If at the end, all you are doing to add context is to append the text to "prompt" or "system" message, then why do you need the overlays?

Because we need to keep track of text regions that have been added to the context.  If the regions change, either because the user or the LLM modifies them, we need to update the context.

Without this `gptel-add` is meaningful for the next request only.  Adding context is less like appending text to the system message, and more like adding a promise or a function that will run `(concat (buffer-substring from to) system-message)` when sending the request.

> And why can't they be made on the fly from gptel-context--alist and not the other way around (making overlays first and keeping a record in the alist)?

We need to indicate to the user (via the coloration) that the text region is part of gptel's context.
  
What other information can you place in `gptel-context--alist` to track regions of text?

----

This is an advanced version of the age-old workspaces problem in Emacs.  There are so many solutions to extending the project metaphor to the UI (perspective, persp, project-tab-groups, beframe, ...) because by its nature, Emacs does not lend itself well to isolated grouping like this.  You have to work hard to force-fit a solution.  I haven't had success getting any of these to stick for over a couple of months because scopes work naturally in Emacs only if they are (i) global or (ii) buffer-local.

And the above issues are the tip of the iceberg.  When you sit down and try to implement this feature you quickly realize you simply can't cover all your bases in a reasonable, visible and maintainable way.

That doesn't mean this can't be done.  But it requires the current context system to be redesigned from scratch, possibly with a buffer-local, not project/workspace-local approach.

**Comment by armindarvish** [comment]
But I think there are ways to go about it. The workflow from the user side would need to be adjusted but I feel like at the end it would make more sense for the user as well.

Here are some ideas: 

> You go to buffer (B), select a region and add it to the "context" with gptel-add.
> 
> Which buffer's value of gptel-context--alist should be updated?

This can be much simpler. First, the user can do the `(gptel-add)` in buffer C and be asked if s(he) wants to add region from another buffer. Alternatively, s(he) can run `add-gptel` in buffer B, and have the chat in buffer C inherit from buffer B (by an orthogonal function that copies the context from another buffer). 


> According to this we would do something like the following: when adding a region, buffer or file to the context in a buffer's local value of gptel-context--alist, find all buffers belonging to that project, then update gptel-context--alist in all of them.
>
>    Now suppose you open some more files from that project, and want to run an operation with gptel (rewriting a region, say) in one of these files. gptel-context--alist in this project buffer has not been updated so we are out of sync. The only way around this is to add to find-file-hook -- only that's not enough! What about non-file project buffers that might be spawned, like compilation or output buffers? So we'd have to register a handler in buffer-list-update-hook instead. This is generally a bad idea. This also doesn't cover all buffers, because there are special buffers whose text you may want to add to the context, but don't run buffer-list-update-hook when they are spawned/killed.


It seems to me that you are assuming that the context has to be in place before running `gptel-send`, but why can't it be dynamically calculated before sending the prompt to the backend. I can see a scenario where a customization variable `gptel-context-inherit-from-project` is set to `t` and that means any time a prompt is sent to the backend the context is created by looking at all the project buffers. So no matter when you opened the file and in which order, it finds all the relevant context for each buffer every time. Maybe you have some concern about speed or performance, etc. but getting the `buffer-substring` form some markers, or file contents and appending them to the prompt or system should be pretty fast in Emacs anyway.  

>   Suppose you run gptel-send from the transient menu from a utility buffer (like *Messages* or *scratch*) that's not associated with any project. The context is now inaccessible. This problem gets worse when you want to add a non-project (or non-generic-group) file to the project's context from dired, as opposed to from the chat buffer. The dired directory will not be a project directory.

In case of transient menu, you can perhaps look up the chat buffer that opened the transient menu associated with it and use the context from there. For *Messages*, *scratch*, etc. you can either manually run `add-gptel` and add context (optionally from other buffers), or with some convenient function like `gptel--context-copy-from-another-buffer` or `gptel--context-append-from-another-buffer`, `gptel--context-inherit-from-another-buffer`, `gptel--context-inherit-from-project`, etc. add context from other buffer(s) or project(s).

>   It's now a mystery which context set an overlay belongs to, or what the active context--alist is at any time. The UI is ambiguous now -- you can have two buffers side by side with visible context overlays and have no idea if one, both or none of the regions will be sent with your next request.

Here is one fundamental piece that I do not understand. What are the overlays for? If you already have something like `'((#<buffer gptel-tools-test> 1 to 2000))` or anything similar that essentially holds the information about the buffer name and markers for the region in the `context-alist` you already have all the information you need for what text to add to the context. Why do you need to add overlays to the buffers?
Still assuming that you absolutely need the overlays because you are worried about text changing in the buffer and invalidating the marker positions, or something else I haven't thought of yet, then you can still keep track of each context by adding text-properties or other overlays that carry the context name, which in this case can be the name of the buffer that the context belongs to since in this case, the context would be buffer-local. 
Also the active context is always the info stored in `gptel--context-alist` in the buffer that sends the prompt. You should not need to look at the overlays in the buffers containing the info, but at `gptel--context-alist` and if you want to see the info, there can be other convenient functions that jump to the regions stored in `gptel--context-alist`.  


By the way, in the scenario you described, most of the problems you mention still exist because one has to discard and rebuild the context from scratch when switching contexts (switching projects, buffers, etc.) and it is still very ambiguous as well because one has to remember to check the global context before calling `gptel-send` in any random buffer and in some cases that would mean discard the whole context, send a one of prompt, then later rebuild the context again.

**Comment by metachip** [comment]
@karthink The core issue seems to be that overlays, while excellent for visual feedback, create constraints when managing multiple contexts because they exist in the buffer space, not context space. This leads to potential conflicts and ambiguity when multiple contexts need to reference the same or overlapping regions.

How about separating the storage of context information from its visual representation. Here's the concept:

1) Design Change:
   + Store context information independently of buffer metadata
   + Generate visual indicators (overlays) only when needed
   + Support for buffer-local context storage
   + Clear hierarchy for context selection
2) Key Benefits:
   + No overlay conflicts between contexts
   + Easier context persistence
   + Clear separation between storage and display
   + Simpler buffer-local context management
3) Context Selection Logic:
   + Buffer-local context takes precedence if present
   + For multiple named contexts, prompt user to select
   + Fall back to global context if no other applies
   + Selection happens only at send time
4) Storage and Persistence:
   + Context data stored in a central registry
   + Buffer-local variables for per-buffer settings
   + File-local variables for persistence
   + File or heading properties for org mode files
   + Project-level context storage possible
5) Visual Feedback:
   + Temporary overlay generation when needed
   + Clear visual indication of active context
   + No permanent buffer modifications
   + Clean separation of display from data

This approach would maintain the simplicity of the current implementation while adding flexibility for multiple context use cases. It also makes it easier to save and restore contexts with buffers, which addresses one of the original motivations for this feature request. The key insight would be we don't use permanent overlays to track regions - we can store that information separately as part of the meta data as I've suggest and generate the visual feedback when needed, particularly at send time or when inspecting contexts. This could provide a path forward that doesn't compromise the current simple usage while enabling more sophisticated context management for those who need it.

Thoughts?

**Comment by karthink** [comment]
> The core issue seems to be that overlays, while excellent for visual feedback, create constraints when managing multiple contexts because they exist in the buffer space, not context space.

Overlays are used to track regions in buffers and for visual feedback.  Removing the visual feedback doesn't help because you still need to track regions.  It doesn't matter that we use overlays for this.  If you replaced each overlay with a pair of markers you still have the problem that they "exist in the buffer space", as you put it.

> This leads to potential conflicts and ambiguity when multiple contexts need to reference the same or overlapping regions.

This is not an issue, you can have overlapping overlays.

> How about separating the storage of context information from its visual representation.

Doesn't help, as I explained above.

The rest of the suggestions have the same problems (complexity and ambiguity) as the original proposal in this thread.

Notice that @ultronozm's [prototype for buffer-local context above](https://github.com/karthink/gptel/issues/475#issuecomment-2590760043) does not include bullfer-local region contexts.  As they put it:

>  We disallow "buffer-local context overlays", which would probably be confusing

Try implementing it (with or without overlays) and you'll see why.

**Comment by metachip** [comment]
Looking back through this thread, I think @dwrz and @nlevnaut's suggestions point to the simplest workable solution, with an additional refinement for persistence:

1) Keep the current single-context mechanism exactly as is
2) Add ability to:
   + Save current context with a name
   + Switch to a previously saved context
   + List available contexts
   + Delete saved contexts

This maintains GPTEL's simplicity while solving the core need - managing different contexts for different tasks. Key aspects:

+ Only one active context at any time
+ Complete context switch (no mixing/inheriting)
+ Clear visual feedback remains unchanged
+ No ambiguity about what's in context

Additionally, for buffer-specific persistence, we could save context information similar to how GPTEL already handles model selection and bounds. The key differences would be:

1) Store the complete context definition, not just a name
2) Only persist file-based information:
   + Full file paths for file contexts
   + File paths + positions for regions from file-backed buffers
   + Skip regions from temporary/unsaved buffers

Example serialized format:
``` org
:PROPERTIES:
:GPTEL_MODEL: claude-3-sonnet
:GPTEL_CONTEXT: (:files ("/path/to/file1" "/path/to/file2") :regions (("/path/to/file3" 1 100) ("/path/to/file4" 200 300)))
:END:
```

When gptel-send is called:
1) Check for stored context definition
2) Restore that exact context if found
3) Otherwise use the current global context

This approach provides both:
+ Simple global context management through named contexts
+ Robust per-buffer context persistence without dependencies on buffer state

Would one or both of these ideas in combination be worth pursuing?

**Comment by karthink** [comment]
> It seems to me that you are assuming that the context has to be in place before running gptel-send, but why can't it be dynamically calculated before sending the prompt to the backend. I can see a scenario where a customization variable gptel-context-inherit-from-project is set to t and that means any time a prompt is sent to the backend the context is created by looking at all the project buffers. So no matter when you opened the file and in which order, it finds all the relevant context for each buffer every time. 

There are plans to integrate dynamic context in a customizable way (from a RAG source, LSP etc).  Writing these sources is out of scope for gptel, but we do want to provide an easy way to plug them in and toggle them.  There are several issues and discussions about this if you're interested.  (You can also easily add dynamic context yourself; all you need is to run an elisp function and append the result to the prompt or system message.)

But that's not what this feature is about.  You need a way to manually and unambiguously specify context regions, without relying on some algorithm.  This is the purpose of gptel-context.

> Maybe you have some concern about speed or performance, etc. but getting the buffer-substring form some markers, or file contents and appending them to the prompt or system should be pretty fast in Emacs anyway.

This is what gptel-context currently does, so there is some misunderstanding here.  Have you looked at the code in gptel-context?

There is also no functional difference between an overlay and a pair of markers.

> This can be much simpler. First, the user can do the (gptel-add) in buffer C and be asked if s(he) wants to add region from another buffer. Alternatively, s(he) can run add-gptel in buffer B, and have the chat in buffer C inherit from buffer B (by an orthogonal function that copies the context from another buffer).

Sorry, I cannot explain any better why this won't work -- and if it does work it's inconsistent and confusing.  @metachip put it well above when they said "[context regions] exist in the buffer space, not context space".

You can try to implement it and see for yourself.

> In case of transient menu, you can perhaps look up the chat buffer that opened the transient menu associated with it and use the context from there. 

This is how all transient menus work, they are automatically scoped to the buffer they are invoked from.  So this has no (positive or negative) effect on the issues with buffer-local context.

> For Messages, scratch, etc. you can either manually run add-gptel and add context (optionally from other buffers), 

The problem is adding partial contents of these buffers to the context of other buffers, especially ones it doesn't share a group with.

> or with some convenient function like gptel--context-copy-from-another-buffer or gptel--context-append-from-another-buffer, gptel--context-inherit-from-another-buffer, gptel--context-inherit-from-project, etc. add context from other buffer(s) or project(s).

Again, I can't explain any better why this is going to be confusing to most users.

> Here is one fundamental piece that I do not understand. What are the overlays for? If you already have something like '((#<buffer gptel-tools-test> 1 to 2000)) or anything similar that essentially holds the information about the buffer name and markers for the region in the context-alist you already have all the information you need for what text to add to the context. Why do you need to add overlays to the buffers?

Forget about the overlays for a moment, I think the visual aspect is throwing us off-course.  You can just picture them as markers, the same issues apply.

> Also the active context is always the info stored in gptel--context-alist in the buffer that sends the prompt. You should not need to look at the overlays in the buffers containing the info, but at gptel--context-alist and if you want to see the info, there can be other convenient functions that jump to the regions stored in gptel--context-alist.

You don't need to look at the overlays.  You can jump from the context inspection buffer (which is basically a UI for inspecting `gptel-context--alist`) to the source by pressing `RET`.

I don't understand why you believe visual feedback is the problem -- it's not, the problem is the markers.  We just get visual feedback for free if we use overlays instead of markers.  Independent of this, I also think visual feedback is important.

**Comment by ultronozm** [comment]
Just to add a bit on this:

> Notice that [@ultronozm](https://github.com/ultronozm)'s [prototype for buffer-local context above](https://github.com/karthink/gptel/issues/475#issuecomment-2590760043) does not include bullfer-local region contexts. As they put it:
> 
> > We disallow "buffer-local context overlays", which would probably be confusing

The substitute for overlays in that proposal is to do `clone-indirect-buffer` and `narrow-to-region` on the context region.  This gives a fresh buffer that serves as a "window" into the context region.  Crucially, this new buffer has a name, so we can refer to it, with completion, when we want to specify local context elsewhere.

There's then the separate UI question of optimizing the ergonomics (e.g., by providing a helper function that clones, narrows and renames to something like *gptel-region-#*, which is close to what I do in practice).

Personally, I prefer this approach even for global context because it makes it easy to tweak the boundaries by widening/narrowing.

**Comment by metachip** [comment]
> Personally, I prefer this approach even for global context because it makes it easy to tweak the boundaries by widening/narrowing.

I agree. The need for a context buffer, that shows all the context regions together, even if enhanced to be collapsible like Magit does, does not make as much sense as indirect buffers, for viewing, managing and editing (text) context. The benefit of indirect buffers is that "everything in context is available via a buffer" and each buffer has a name which makes it all explicit.

**Comment by armindarvish** [comment]
@karthink Ok, let me experiment with the buffer-local context and see what I can come up with. My focus on the overlays (visual feedback) was merely a response to what you said; otherwise, it appears that we are saying the same thing: A set of markers would do the job just as well. 

The part I don't understand is why you think that would be confusing to users! The buffer space v.s. context space is not critical because the user does not need to see any of that. Each buffer has its own context (a set of markers to locations/regions). Any request sent from this buffer will append those regions as context. Other buffers have their own context stored in their own `context-alist` variable. The fact that in one buffer, there can be markers that do not belong to the context of that buffer itself is not important because the "single source of truth" for the context will be the buffer-local `context-alist` variable and not the markers (overlays, etc.) in the buffer itself. This is already the case with the global scope of `context-alist` anyway. The buffer-local version would just expand that to allow having different contexts in different buffers, otherwise everything else would feel the same, at least to the end user.

As soon as I find some time, I'll whip up something to see what is the confusing part you are all concerned about.

**Comment by armindarvish** [comment]
One quick question for the mean time, though. 

What is the best way to ignore context in one-off requests? 

This is especially important for privacy purposes. If I add some data to the context in a chat that I send to a local server, how do I make sure that this context does not get appended to requests that I may send later to remote servers? For example, consider how I use gptel in consult-omni. I am sending requests to gptel on the fly and even changing the backend/models/etc. from the minibuffer input, and I would probably not want to add the global context to every one of those requests. Would a let-binding of `gptel-use-context` to `nil` around the request suffice, or do I need to also do other things like setting `gptel--context-alist` to nil?

**Comment by karthink** [comment]
> One quick question for the mean time, though.
> 
> What is the best way to ignore context in one-off requests?
> 
> This is especially important for privacy purposes. If I add some data to the context in a chat that I send to a local server, how do I make sure that this context does not get appended to requests that I may send later to remote servers? For example, consider how I use gptel in consult-omni. I am sending requests to gptel on the fly and even changing the backend/models/etc. from the minibuffer input, and I would probably not want to add the global context to every one of those requests. Would a let-binding of `gptel-use-context` to `nil` around the request suffice, or do I need to also do other things like setting `gptel--context-alist` to nil?

Via elisp: let-binding `gptel-use-context` to `nil` is all you need.
Interactively: Turn off the context globally, for requests in this buffer, or for the next request only:

![Image](https://github.com/user-attachments/assets/6a64eb18-e829-46f4-b5ea-9a6fbdde0c9a)

**Comment by metachip** [comment]
# Context by links instead

> What you're describing here is an alternative syntax to the link method you proposed in #481.

Yep. And I prefer that method if it comes to a choice because it's explicit, declarative and can be used where and whenever one wants to include context, whatever that context may be. I also agree and understand the issue of the alignment of region references to buffer contents when the buffer is modified. Context links would not support regions, given the org link syntax does not allow them to be specified without a link name "hack" of the sort I proposed in #481.

> I don't follow. You don't need any special affordance for viewing or editing context. You don't even have to think about it -- you just edit the buffer the context region is in.

Sure but finding the buffers in which referenced context regions are located requires one to traverse the context buffer (to open the buffer) or one needs to "remember where". I would prefer the context buffer to simply lists these references and not the context itself. One would of course be forced to select a reference to open a buffer to view the context but I think this is better anway - I understand and like that editing context is done in the buffer that contains the context but...

> The advantage of a narrowed indirect buffer is that your reference to the context is a buffer, which is easier to refer to and serialize than an overlay.

I think indirect buffers are a better idea from a UI PoV than overlays anyway. For the reason you cite and for simplicity. Overlays in buffers can be difficult to view and use. Indirect buffers are completely straight forward by comparison from a UI PoV.

> But it doesn't solve the problem of adding a region to buffer B's context by running gptel-add in buffer A.

Agreed. My most recent suggestion was that there should only be one (selected global) context (at a time). I'm personally happy to abandon the idea of multiple concurrent contexts altogether. One could still allow the (global) context to be saved as a local variable or property in a buffer, and restored to the global context from that buffer upon command. It would provide a transparent way to manage the global context and allow context to be saved and restored with (gptel-mode) buffers to/from files going some way to offer a solution to this thread.

In my case, all this could be largely moot if #481 was available as an alternative.

**Comment by algal** [comment]
@metachip I was wondering, did you get link-based contexts working in any private branches?

I would also love some way to define concurrent, isolated, project-local contexts. But in a pinch, savable/restorable contexts see like a reasonable a good substitute, and saving context definitions via links seems like a reasonable solution or at least a good stopgap.  

I wrote a couple helper functions to try this idea( https://github.com/algal/alg-gptel-helpers ), a function which simply saves the context as a lisp form, and one which restores it by executing the form, but I'm hoping there is something better out there.

**Comment by metachip** [comment]
Not yet, although it is on the roadmap. My preferred choice is to use links to create context. I'll have a look at your helper functions when I get a chance. Suffice to say I generally work in org mode buffers, so heading properties is another way in which one might do this.

**Comment by karthink** [comment]
I have a prototype of context via links on my machine with many edge cases remaining to be fixed.  I'm sitting on it for now as I need to pretty much rewrite gptel-context to be more modular.

**Comment by dwrz** [comment]
I've lost the thread here a bit, so my apologies if this has already been discussed -- would it be possible to use the `Scope` parameter (bound to `=`) for contexts?

**Comment by karthink** [comment]
> I've lost the thread here a bit, so my apologies if this has already been discussed -- would it be possible to use the `Scope` parameter (bound to `=`) for contexts? 

Unfortunately the scope does not apply to contexts because context chunks are necessarily global, they're overlays in other buffers.  The last few messages above this one are about providing files as context via Org links in chat buffers.  (This is currently how gptel supplies images to vision-capable models.)

The scope setting affects all options below it in its column -- hope that's somewhat clear from its placement in the menu.

**Comment by ultronozm** [comment]
@algal 

> I would also love some way to define concurrent, isolated, project-local contexts.

For the special case of "chat files" (e.g., an "ai.org" or "ai.md" file at the top level of your project), you could do this using the prototype described at https://github.com/karthink/gptel/issues/475#issuecomment-2590760043, which saves "buffer-local context" in file local variables.

**Comment by krvkir** [comment]
Hi there! 

I'm surprised that no one suggested **to use tab-bar tabs as context scopes.** Tabs seem to be a natural way to organize work, and they opt out all the corner cases you discussed above and don't hinder user when they work with gptel. Tabs are part of stock Emacs so they create no extra dependencies.

I made a small package which binds gptel context to active tabs and silently switches contexts when you switch tabs:

https://github.com/krvkir/gptel-tab/

It's just born and not well-tested yet, but works for me, saving manual dumping and restoring contexts and not cluttering the way I use gptel.

I understand that not all people use tabs to separate their scopes, but for those who do — binding context to tabs seems a very natural thing.

**Comment by karthink** [comment]
From @kamomehettapi in https://github.com/karthink/gptel/issues/839#issuecomment-2877335656:

> Understood but what if I dont care about buffer overlays but only a set of filepath strings I can swap between. I never use buffer regions for context only filepaths because as you said it gets very confusing very fast especially when whole-file updates are made
> 
> It sounds like most complexity about context is looking at buffer regions that might be updated later and those I can understand are hard to save and restore. But to your three questions on [#475](https://github.com/karthink/gptel/issues/475), these are what I think for my needs
> 
> > Suppose you visit some buffer, select a region and run gpel-add to add it to the context.
> > Which gptel buffer's context (context set) should this be added to?
> 
> It adds to current context set but when switching it is not saved
> 
> > How about when there is no chat buffer? gptel is usable in any buffer, and most of the time I don't use a chat buffer.
> > Same question as 1, but for when you run gptel-add in a dired buffer to add a file.
> 
> It adds to current context set but when switching it is not saved
> 
> > Suppose you run gptel-add-file, either from the transient menu or via M-x. This command can be run in any buffer. Which context set should it be added to?
> 
> It is saved to current context set and restorable

**Comment by karthink** [comment]
> Understood but what if I dont care about buffer overlays but only a set of filepath strings I can swap between

I have to try to make gptel work consistently.  Users don't know (and shouldn't have to know or care) about the limitations of using overlays to track buffers/regions etc.  From their perspective, if a context set can be switched to, it should include whatever they added to it.  Otherwise the feature cannot be trusted.

If you are aware of the limitations and interested in managing lists of files, you can use the code for context set switching I posted further up in this thread.

> > Suppose you visit some buffer, select a region and run gpel-add to add it to the context.
> > Which gptel buffer's context (context set) should this be added to?
> 
> It adds to current context set but when switching it is not saved

It _is_ saved, actually.  It's just not serializable, so it can't be defined/saved in your config or as Org properties as a pre-made context set and used across Emacs sessions.

> > How about when there is no chat buffer? gptel is usable in any buffer, and most of the time I don't use a chat buffer.
> > Same question as 1, but for when you run gptel-add in a dired buffer to add a file.
> 
> It adds to current context set but when switching it is not saved

It is saved when you switch contexts.

> > Suppose you run gptel-add-file, either from the transient menu or via M-x. This command can be run in any buffer. Which context set should it be added to?
> 
> It is saved to current context set and restorable

Global context sets work fine as far as saving everything (including overlays) and switching between them within an Emacs session is concerned.  The problems with it are different:
1. They would work differently from how every other setting in gptel works.  All settings can be set globally, buffer-locally or for the next request only.
2. You cannot tell, from looking at a context overlay, which context set it belongs to.  Having tried it, this gets quite confusing quickly.
3. Even from a chat buffer, you have to keep checking the transient menu to ensure that the context set that is active is the one you think is active.  This is not so for any other setting, since they can be set buffer-locally once, and it will not change if you change it somewhere else in Emacs.

Unfortunately I don't see a good solution here, only band-aids.  Context management in gptel needs to be redesigned from the ground up.

**Comment by karthink** [comment]
1. gptel now follows "standalone" links to plain text files and includes their contents in the request.  This is an easy,  intuitive and persistent way to get buffer-local context -- just add links to the files you want to include.  See https://github.com/karthink/gptel/issues/481#issuecomment-2918553606 for details.
2. I have two questions about this design, see https://github.com/karthink/gptel/issues/481#issuecomment-2918562122.

**Comment by dwrz** [comment]
I want to experiment a bit more, but I think link-based context probably covers the base case for me. 🙏🏽

The only other idea that still lingers with me is to expose the context interface to modification via Emacs LISP, if it isn't already. My familiarity with is limited, but could I call the commands in `gptel-context.el` to edit the context programmatically? That way, I could write my own functions to switch among typical context sets.

**Comment by algal** [comment]
This also seems to cover the bases for me. 🙏 .

I am so excited by this feature! It solves my biggest pain point with `gptel`. Now I can conduct simultaneous chats, each with their own context, and I can save and load those chats across emacs sessions.

I started using this the moment it came out. I'm sure I'll notice more ergonomic points as I use it more, but here's one critical observation.

The one significant friction I've noticed right away is that now I can't use `gptel-add-file` in the way I used to. I used to switch to a buffer visiting a file, run `gptel-add-file`, add the file, and then switch back to my gptel chat buffer. Now I am obliged to use `org-insert-link` within the chat buffer, to specify the `file:` scheme, and then to browse for the file. Adding a file to context is such a frequent action that this friction is noticeable.

To help, I wrote a function, `alg/org-kill-file-link` (https://github.com/algal/alg-gptel-helpers/blob/b24d26165ea52829ffeab1268aa7532f2e595ea5/alg-gptel-helpers.el#L11), function which simply saves into the kill buffer a valid org link to the current file, or to the file at the current line in a dired or magic buffer. This lets me go back to my old flow, where I can navigate through emacs and to build a link which I will then yank back into the gptel chat buffer. I find this very handy; I'm not sure how many other people would need it.

**Comment by karthink** [comment]
> The only other idea that still lingers with me is to expose the context
> interface to modification via Emacs LISP, if it isn't already. My familiarity is
> limited, but could I call the commands in `gptel-context.el` to edit the context
> programmatically?

You can add files to `gptel-context--alist` via elisp, but this is not officially supported.  Basically it's
``` emacs-lisp
(push '("/path/to/file") gptel-context--alist)
```

You can set `gptel-context--alist` buffer-locally, with something like this:

``` emacs-lisp
(setq-local gptel-context--alist (copy-alist gptel-context--alist))
```

Then adding to it programmatically will add it to this buffer's value of the context instead of the global one.

Again, the ergonomics of doing this aren't great and this interface is not guaranteed to be stable.  The Org/Markdown link method is more user-friendly.

**Comment by ahmed-shariff** [comment]
This stemmed from our conversation on temporary presets to store context (https://github.com/karthink/gptel/issues/542#issuecomment-2869497351). Probably also related to https://github.com/karthink/gptel/issues/880
This is also related to dumping and recreating contexts.

What I am doing here is creating persistent files with contexts that would be similar to the `*gptel-kill-ring-context*` buffer created by `gptel-context-add-current-kill`. Here I am adding the highlighted region to a file. On the fly, I am also creating presets for each "persistent context file" which sets the `gptel-context--alist`:
```emacs-lisp
(defvar gptel-persistent-context-dir "~/.emacs.d/.cache/gptel-temp-context")

(dolist (context-file (directory-files gptel-persistent-context-dir
                                       nil
                                       directory-files-no-dot-files-regexp))
  (gptel-make-preset (intern (f-base context-file))
    :description (format "Use persistent context %s" (f-base context-file))
    :context--alist `((,context-file))))

(defun my/gptel--get-persistent-context ()
  (completing-read "Context name: "
                   (-map #'f-base
                         (directory-files gptel-persistent-context-dir
                                          nil
                                          directory-files-no-dot-files-regexp))))

(defun my/gptel--get-context-file (context-name)
  (file-truename
   (expand-file-name context-name
                     gptel-persistent-context-dir)))

(defun my/gptel-add-region-to-persistent-context (context-name)
  "Adding the region to the persistent context CONTEXT-NAME."
  (interactive (list (my/gptel--get-persistent-context)))
  (if (not (use-region-p))
      (user-error "Nor regions highligted")
    (let ((content (buffer-substring (region-beginning) (region-end)))
          (context-file (my/gptel--get-context-file context-name)))
      (make-directory gptel-persistent-context-dir t)
      (with-temp-file context-file
        (when (file-exists-p context-file)
          (insert-file-contents context-file))
        (goto-char (point-max))
        (unless (bobp)
          (insert "\n---------\n"))
        (insert content))
      (gptel-make-preset (intern context-name)
        :description (format "Use persistent context %s" context-name)
        :context--alist `((,context-file))))))

(defun my/gptel-clear-persistent-context (context-name)
  "Clear persistent context CONTEXT-NAME."
  (interactive (list (my/gptel--get-persistent-context)))
  (let ((context-file (my/gptel--get-context-file context-name)))
    (when (y-or-n-p (format "Clear the context %s" context-name))
      (with-temp-file context-file))))

(defun my/gptel-visit-persistent-context (context-name)
  "Visit persistent context CONTEXT-NAME."
  (interactive (list (my/gptel--get-persistent-context)))
  (let ((context-file (my/gptel--get-context-file context-name)))
    (find-file context-file)))

(defun my/gptel-delete-persistent-context (context-name)
  "Delete context CONTEXT-NAME."
  (interactive (list (my/gptel--get-persistent-context)))
  (let ((context-file (my/gptel--get-context-file context-name)))
    (when (y-or-n-p (format "Delete the context file for context %s" context-name))
      (delete-file context-file)
      (setf (alist-get (intern context-name) gptel--known-presets nil 'remove) nil))))

;; Because why not!
(transient-append-suffix 'gptel-menu '(0 -1)
  [""
   "Persistent context"
   ("pa" "Append region" my/gptel-add-region-to-persistent-context :transient t)
   ("pc" "Clear" my/gptel-clear-persistent-context :transient t)
   ("pv" "Visit" my/gptel-visit-persistent-context :transient nil)
   ("pd" "Delete" my/gptel-delete-persistent-context :transient t)])
```

This works well for my use-cases where I have a lot of task-specific contexts I use across multiple sessions. But it doesn't handle cases like changing content in a region added to a context.

**Comment by arichiardi** [comment]
Just a thought on this. Now that we have presets, would it be a good idea to attach context (and load it on activation to them).
The two may or may not be linked for experienced users but probably for novices it can be handy and saves much of the "rebuilding" time.

**Comment by 11111000000** [comment]
https://github.com/11111000000/pro/blob/main/libs/gptel-context-store.el

**Comment by madorian** [comment]
> https://github.com/11111000000/pro/blob/main/libs/gptel-context-store.el

https://github.com/11111000000/pro/blob/5b62f8bd608cd4da200677d9bd0f784f590361ab/gptel-context-store.el#L4

**Comment by karthink** [comment]
Per-buffer gptel context specification is now available.

To use per-buffer contexts, you can modify the new `gptel-context` user option buffer-locally.  Please see its documentation.

One way to use it would be to make `gptel-context` a buffer-local variable via `(make-local-variable 'gptel-context)`.  

Then you can add/remove buffer-local context from `gptel-menu` with the `-b` and `-f` options.

You could also just set the context directly, via
```emacs-lisp
(setq-local gptel-context (list "/path/to/file1" "./file2.png" (get-buffer "*scratch*")))
```

You can set it via a preset, as
```emacs-lisp
(gptel-make-preset 'with-docs
  :description "Include Readme files in the current directory with gptel requests"
  :context '("./README.md" "./README.org"))
```

In programmatic use, you can let-bind `gptel-context` around a call to `gptel-send`:
```emacs-lisp
(let ((gptel-context '("/path/to/file1" "/path/to/file2")))
  (gptel-send))
```

If you use buffer-local context, gptel's global context (i.e. the global value of `gptel-context`) will be ignored.

--------

(Note that you cannot set buffer-local context in one buffer by calling interactive functions like `gptel-add` in another buffer... because how would it know which buffer's context you want to add this region to?)

**Comment by ahmed-shariff** [comment]
This is very nice. Thank you very much. Can we add `:context` as a key to `gptel--save-presets`? That also would solve switching between contexts? Though overlays would need to be handled appropriately.

**Comment by karthink** [comment]
> Can we add `:context` as a key to `gptel--save-presets`?

@ahmed-shariff it doesn't as of now.  I expect it will create more confusion than it will provide utility, and not just because of the overlays. You may not realize that you're baking in the current gptel-context into a preset when you save it.

Also, in retrospect the use of overlays for context has proven to be bad design :(

**Comment by ahmed-shariff** [comment]
Would it make sense to have a similar pattern to saving presets and using them, but specific to contexts? Call it "context bins" or "named context"? With overlays, the face changes if it is not in the current bin (maybe just remove the face altogether). C-u on any of the gptel add context functions may prompt for the bin, otherwise add to the current context. 

Very similar designs have been discussed before, in this thread and others. Does the current implementation for gptel-context have the same issues with a design like this?

**Comment by karthink** [comment]
> Does the current implementation for gptel-context have the same issues with a design like this?

What issues are you referring to? 

All I did was to rename the internal variable `gptel-context--alist` to `gptel-context` and handle some edge cases, so the design of context handling hasn't really changed. 

Any of the previous implementations of context bins in this thread can do what you ask.  I don't want to add "context sets" to gptel as a new UI concept, but perhaps some customizable way to include it when saving a preset might be acceptable.

**Comment by ahmed-shariff** [comment]
> What issues are you referring to?

I was referring to these comments: https://github.com/karthink/gptel/issues/475#issuecomment-2877672264

> perhaps some customizable way to include it when saving a preset might be acceptable.

That would be an option to customize the key-value pairs that `gptel--save-preset` generates? That also could be a nice solution for now.

**Comment by karthink** [comment]
> That would be an option to customize the key-value pairs that `gptel--save-preset` generates? That also could be a nice solution for now.

Yup I'll add this.  Do you have a suggestion for a better name for this variable?

``` emacs-lisp
(defcustom gptel-preset-default-options
  '( gptel-backend gptel-model gptel--system-message gptel-tools
     gptel-stream gptel-temperature gptel-max-tokens
     gptel-use-context gptel-track-media gptel-include-reasoning)
  "gptel settings whose values should be saved into presets.

This is a list of gptel options (variables) whose values will be saved
to a preset when creating a new preset interactively."
  :type '(repeat symbol)
  :group 'gptel)
```

Incidentally, I tried including `gptel-context` in this list and it immediately became very confusing when switching between presets.  I think some other interactive way of setting the context locally is required.

**Comment by ahmed-shariff** [comment]
> Do you have a suggestion for a better name for this variable?

Should it be `gptel-preset-saved-options` or `gptel-preset-options-to-save`?

> it immediately became very confusing when switching between presets

You are right. I was playing around with adding a custom UI to some of this and very quickly started running into some interesting cases. Here's what I have for now - it fits some of my current use cases, but still has some issues. For example, adding a new region that already has a context overlay and how deleted overlays are to be handled:

```emacs-lisp
(defun amsha/gptel--save-context-as-preset (name &optional description)
  "Save a preset with only the context.

Same as `gptel--save-preset', only save context."
  (interactive
   (list (intern (completing-read "Save gptel settings to (existing or new) preset: "
                                  gptel--known-presets))
         (read-string "Description (optional): ")))
  (let ((preset-code
         `(gptel-make-preset ',name
           :description ,(when (and description
                                (not (string-blank-p description)))
                           description)
           :context ',gptel-context)))
    (kill-new (pp-to-string preset-code))
    (eval preset-code)
    (message "Preset %s saved. (Lisp expression for preset saved to kill-ring)"
             (propertize (symbol-name name) 'face 'highlight))))

(defun amsha/gptel-context-remove-all-except-overlays (&optional verbose)
  "Remove all gptel context.

If VERBOSE is non-nil, ask for confirmation and message
afterwards.
Copied from `gptel-context-remove-all'."
  (interactive (list t))
  (if (null gptel-context)
      (when verbose (message "No gptel context sources to remove."))
    (when (or (not verbose) (y-or-n-p "Remove all context except overlays? "))
      (cl-loop
       for context in gptel-context
       for (source . ovs) = (ensure-list context)
       unless (cl-every #'overlayp ovs) do ; unless buffers and buffer regions
       (gptel-context-remove source) ;files or other types
       finally do (setq gptel-context nil)))
    (when verbose (message "Removed all gptel context sources except overlays."))))

(defun amsha/remove-context-at-point ()
  (interactive)
  (gptel-context-remove))

(transient-suffix-put 'gptel-menu '(0 1 6) :key "-D")
(transient-insert-suffix 'gptel-menu '(0 1 6)
  '("-d" "Remove non overlays" amsha/gptel-context-remove-all-except-overlays
    :if (lambda () gptel-context)
    :transient t))

(transient-insert-suffix 'gptel-menu '(0 1 7)
  '("/r" "Remove context at point" amsha/remove-context-at-point
    :if (lambda () (or (region-active-p) (gptel-context--at-point)))
    :transient t))

(transient-append-suffix 'gptel--preset '(0 0 -1)
  '("/C" "Save current context as new preset" amsha/gptel--save-context-as-preset :if (lambda () gptel-context)))
```

**Comment by 11111000000** [comment]
https://github.com/11111000000/context-navigator

I create this to rule context
